### PR TITLE
Dzhus/organization(provider) update

### PR DESCRIPTION
**Pull request by VyacheslavDzhus**

### CODE REVIEW

File: OutOfSchool/OutOfSchool.Tests/ProviderControllerTest.cs [hunk 1]
```diff
@@ -0,0 +1,26 @@
...
+using NUnit.Framework;
+using OutOfSchool.WebApi.Controllers;
+using OutOfSchool.WebApi.Services;
+
+namespace OutOfSchool.Tests
+{
+    [TestFixture]
+    public class Tests
+    {
+        private readonly ILogger<ProviderController> _logger;
```
Comment (dmitrykiev) [inline OutOfSchool/OutOfSchool.Tests/ProviderControllerTest.cs:12]: please do not use underscore naming

File: OutOfSchool/OutOfSchool.DataAccess/Repository/IProviderRepository.cs [hunk 2]
```diff
@@ -5,8 +5,8 @@
 
 namespace OutOfSchool.Services.Repository
 {
-    public interface IOrganizationRepository : IEntityRepository<Organization>
+    public interface IProviderRepository : IEntityRepository<Provider>
     {
-        bool IsUnique(Organization entity);
+        bool IsNotUnique(Provider entity);
```
Comment (dmitrykiev) [inline OutOfSchool/OutOfSchool.DataAccess/Repository/IProviderRepository.cs:10]: the old name 'IsUnique' was better
the rule is to not create negating names
imagine how it will look like '!IsNotUnique' how hard it would be to figure out what is the the result of the operation

File: OutOfSchool/OutOfSchool.WebApi.Tests/Controllers/ProviderControllerTest.cs [hunk 3]
```diff
@@ -0,0 +1,26 @@
+using Microsoft.Extensions.Logging;
```
Comment (dmitrykiev) [inline OutOfSchool/OutOfSchool.WebApi.Tests/Controllers/ProviderControllerTest.cs:1]: looks like we have 2 files ProviderControllerTest.cs 
could you check which one is not needed?

File: OutOfSchool/OutOfSchool.WebApi/Controllers/ProviderController.cs [hunk 4]
```diff
@@ -0,0 +1,125 @@
...
+        /// <summary>
+        /// Initializes a new instance of the <see cref="ProviderController"/> class.
+        /// </summary>
+        /// <param name="providerService">Service for Provider model.</param>
+        public ProviderController(ILogger<ProviderController> logger, IProviderService providerService)
+        {
+            this.logger = logger;
+            this.providerService = providerService;
+        }
+
```
Comment (dmitrykiev) [inline OutOfSchool/OutOfSchool.WebApi/Controllers/ProviderController.cs:30]: remove extra newline

File: OutOfSchool/OutOfSchool.WebApi/Controllers/ProviderController.cs [hunk 5]
```diff
@@ -0,0 +1,125 @@
...
+            if (id == 0)
+            { 
+                return BadRequest("Id cannot be 0.");
+            }
+
+            return Ok(await providerService.GetById(id).ConfigureAwait(false));
+        }
+
+        /// <summary>
+        /// Method for creating new provider.
```
Comment (dmitrykiev) [inline OutOfSchool/OutOfSchool.WebApi/Controllers/ProviderController.cs:59]: Could be just 'Create new provider.'

File: OutOfSchool/OutOfSchool.WebApi/Controllers/ProviderController.cs [hunk 6]
```diff
@@ -0,0 +1,125 @@
...
+                providerDTO.UserId = User.FindFirst("sub")?.Value;
+                ProviderDTO provider = await providerService.Create(providerDTO).ConfigureAwait(false);
+                return CreatedAtAction(
+                    nameof(GetProviderById),
+                    new
+                    {
+                        id = provider.Id,
+                    }, provider);              
+            }
+            catch (Exception ex)
```
Comment (dmitrykiev) [inline OutOfSchool/OutOfSchool.WebApi/Controllers/ProviderController.cs:83]: do not catch base Exception type

File: OutOfSchool/OutOfSchool.WebApi/Controllers/ProviderController.cs (hunk 5 above)
Comment (VyacheslavDzhus) [reply OutOfSchool/OutOfSchool.WebApi/Controllers/ProviderController.cs:59]: How about just CreateProvider ? Other controllers have the same naming.

File: OutOfSchool/OutOfSchool.WebApi/Services/ProviderService.cs [hunk 7]
```diff
@@ -0,0 +1,104 @@
...
+            var provider = dto.ToDomain();
+
+            if (!repository.IsUnique(provider))
+            {
+                throw new ArgumentException("There is already an providerDto with such data");
+            }
+     
+            return CreateInternal(provider);
+        }
+        private async Task<ProviderDto> CreateInternal(Provider provider)
```
Comment (dmitrykiev) [inline OutOfSchool/OutOfSchool.WebApi/Services/ProviderService.cs:46]: move private methods after public methods

File: OutOfSchool/OutOfSchool.DataAccess/Enums/Gender.cs [hunk 8]
```diff
@@ -3,6 +3,6 @@
     public enum Gender
     {
         Male,
-        Female
+        Female,
```
Comment (dmitrykiev) [inline OutOfSchool/OutOfSchool.DataAccess/Enums/Gender.cs:6]: unnecessary change

File: OutOfSchool/OutOfSchool.DataAccess/Models/Teacher.cs [hunk 9]
```diff
@@ -1,42 +1,40 @@
...
         public long TeacherId { get; set; }
-        
+
         [Required(ErrorMessage = "First name is required")]
         [DataType(DataType.Text)]
         [MaxLength(30)]
         [RegularExpression(@"^([^0-9]*)$", ErrorMessage = "First name cannot contains digits")]
-        public string FirstName { get; set; }
-        
+        public string? FirstName { get; set; }
```
Comment (mmmpolishchuk) [inline OutOfSchool/OutOfSchool.DataAccess/Models/Teacher.cs:15]: Can Firstname, Lastname and middlename be nullable?

File: OutOfSchool/OutOfSchool.DataAccess/Models/Workshop.cs [hunk 10]
```diff
@@ -13,20 +14,21 @@ public class Workshop
         [Required(ErrorMessage = "Workshop title is required")]
         [MinLength(1)]
         [MaxLength(60)]
-        public string Title { get; set; }
+        public string? Title { get; set; }
```
Comment (mmmpolishchuk) [inline OutOfSchool/OutOfSchool.DataAccess/Models/Workshop.cs:17]: Can Title be null, as it is required and with minlength (1)?

**Review by mmmpolishchuk — APPROVED** [review]

File: OutOfSchool/OutOfSchool.WebApi/Startup.cs [hunk 11]
```diff
@@ -42,7 +64,8 @@ public void ConfigureServices(IServiceCollection services)
                 });
 
             services.AddCors(confg =>
-                confg.AddPolicy("AllowAll",
+                confg.AddPolicy(
```
Comment (dmitrykiev) [inline OutOfSchool/OutOfSchool.WebApi/Startup.cs:67]: why this change?

File: OutOfSchool/OutOfSchool.WebApi/Startup.cs [hunk 12]
```diff
@@ -55,45 +78,18 @@ public void ConfigureServices(IServiceCollection services)
...
+            services.AddTransient<IProviderRepository, ProviderRepository>();
 
             // Register the Swagger generator, defining 1 or more Swagger documents
             services.AddSwaggerGen();
 
             services.AddAutoMapper(typeof(Startup));
         }
-
-        // This method gets called by the runtime. Use this method to configure the HTTP request pipeline.
-        public void Configure(IApplicationBuilder app, IWebHostEnvironment env)
```
Comment (dmitrykiev) [inline OutOfSchool/OutOfSchool.WebApi/Startup.cs:73]: why this method removed?

File: OutOfSchool/OutOfSchool.WebApi/Services/ProviderService.cs [hunk 13]
```diff
@@ -0,0 +1,103 @@
...
+        {
+            try
+            {
+                await repository
+                    .Delete(await repository.GetById(id).ConfigureAwait(false))
+                    .ConfigureAwait(false);
+            }
+            catch (ArgumentNullException ex)
+            {
+                throw new ArgumentNullException(nameof(id), ex.Message);
```
Comment (dmitrykiev) [inline OutOfSchool/OutOfSchool.WebApi/Services/ProviderService.cs:92]: we throw the same exception and we loose stacktrace here

File: OutOfSchool/OutOfSchool.WebApi/Services/ProviderService.cs [hunk 14]
```diff
@@ -0,0 +1,103 @@
...
+        /// <inheritdoc/>
+        public async Task<ProviderDto> Update(ProviderDto dto)
+        {
+            try
+            {
+                var provider = await repository.Update(dto.ToDomain()).ConfigureAwait(false);
+
+                return provider.ToModel();
+            }
+            catch (Exception ex)
```
Comment (dmitrykiev) [inline OutOfSchool/OutOfSchool.WebApi/Services/ProviderService.cs:75]: do not catch base Exception type

File: OutOfSchool/OutOfSchool.WebApi/Services/ProviderService.cs [hunk 15]
```diff
@@ -0,0 +1,103 @@
...
+             var providers = await repository.GetAll().ConfigureAwait(false);
+
+             return providers.Select(organization => organization.ToModel()).ToList();
+        }
+
+        /// <inheritdoc/>
+        public async Task<ProviderDto> GetById(long id)
+        {
+            var provider = await repository.GetById(id).ConfigureAwait(false);
+            if (provider == null)
```
Comment (dmitrykiev) [inline OutOfSchool/OutOfSchool.WebApi/Services/ProviderService.cs:58]: missing newline

File: OutOfSchool/OutOfSchool.WebApi/Services/ProviderService.cs [hunk 16]
```diff
@@ -0,0 +1,103 @@
...
+            if (dto == null)
+            {
+                throw new ArgumentNullException(nameof(dto), "Provider was null.");
+            }
+
+            var provider = dto.ToDomain();
+
+            if (repository.Exists(provider))
+            {
+                throw new ArgumentException("There is already an providerDto with such data");
```
Comment (dmitrykiev) [inline OutOfSchool/OutOfSchool.WebApi/Services/ProviderService.cs:40]: I believe we should not write 'providerDto' here
it is more clear for the user to see just 'provider'

File: OutOfSchool/OutOfSchool.WebApi/Services/ProviderService.cs [hunk 17]
```diff
@@ -0,0 +1,103 @@
...
+        {
+            this.repository = repository;
+        }
+
+        /// <inheritdoc/>
+        public Task<ProviderDto> Create(ProviderDto dto)
+        {
+            if (dto == null)
+            {
+                throw new ArgumentNullException(nameof(dto), "Provider was null.");
```
Comment (dmitrykiev) [inline OutOfSchool/OutOfSchool.WebApi/Services/ProviderService.cs:33]: Provider was null. -> Provider is null.

File: OutOfSchool/OutOfSchool.WebApi/Models/WorkshopDTO.cs [hunk 18]
```diff
@@ -19,7 +16,8 @@ public class WorkshopDTO
 
         [DataType(DataType.PhoneNumber)]
         [Required(ErrorMessage = "Phone number is required")]
-        [RegularExpression(@"([\d]{9})",
+        [RegularExpression(
```
Comment (dmitrykiev) [inline OutOfSchool/OutOfSchool.WebApi/Models/WorkshopDTO.cs:19]: why this change?

File: OutOfSchool/OutOfSchool.WebApi/Models/ProviderDTO.cs [hunk 19]
```diff
@@ -0,0 +1,112 @@
...
+    {
+        public long Id { get; set; }
+
+        [Required(ErrorMessage = "Title is required")]
+        [DataType(DataType.Text)]
+        [MaxLength(60)]
+        [MinLength(1)]
+        public string Title { get; set; }
+
+        [Required(ErrorMessage = "Title is required")]
```
Comment (dmitrykiev) [inline OutOfSchool/OutOfSchool.WebApi/Models/ProviderDTO.cs:19]: "ShortTitle is required"

File: OutOfSchool/OutOfSchool.WebApi/Models/ProviderDTO.cs [hunk 20]
```diff
@@ -0,0 +1,112 @@
...
+        [MaxLength(60)]
+        [MinLength(1)]
+        public string ShortTitle { get; set; }
+
+        [DataType(DataType.Url)]
+        [MaxLength(30)]
+        public string Website { get; set; } = string.Empty;
+
+        [DataType(DataType.Url)]
+        [MaxLength(30)]
```
Comment (dmitrykiev) [inline OutOfSchool/OutOfSchool.WebApi/Models/ProviderDTO.cs:30]: why 30?

File: OutOfSchool/OutOfSchool.WebApi/Startup.cs (hunk 12 above)
Comment (VyacheslavDzhus) [reply OutOfSchool/OutOfSchool.WebApi/Startup.cs:73]: It is not removed, i put this method over ConfigureServices, because of the StyleCope

File: OutOfSchool/OutOfSchool.WebApi/Models/WorkshopDTO.cs (hunk 18 above)
Comment (VyacheslavDzhus) [reply OutOfSchool/OutOfSchool.WebApi/Models/WorkshopDTO.cs:19]: because of the StyleCope

File: OutOfSchool/OutOfSchool.WebApi/Models/ProviderDTO.cs (hunk 20 above)
Comment (VyacheslavDzhus) [reply OutOfSchool/OutOfSchool.WebApi/Models/ProviderDTO.cs:30]: https://github.com/ita-social-projects/OoS-Backend/issues/54 Max 30 characters

File: OutOfSchool/OutOfSchool.WebApi/Startup.cs (hunk 11 above)
Comment (VyacheslavDzhus) [reply OutOfSchool/OutOfSchool.WebApi/Startup.cs:67]: because of the StyleCope

File: OutOfSchool/OutOfSchool.WebApi/Controllers/ProviderController.cs [hunk 21]
```diff
@@ -0,0 +1,145 @@
...
+        [ProducesResponseType(StatusCodes.Status401Unauthorized)]
+        [ProducesResponseType(StatusCodes.Status200OK)]
+        public async Task<ActionResult> Update(ProviderDto providerDTO)
+        {
+            if (!ModelState.IsValid)
+            {
+                return BadRequest(ModelState);
+            }
+
+            return Ok(await providerService.Update(providerDTO).ConfigureAwait(false));
```
Comment (dmitrykiev) [inline OutOfSchool/OutOfSchool.WebApi/Controllers/ProviderController.cs:120]: please return CreatedAtAction as Mykhailo suggested

File: OutOfSchool/OutOfSchool.WebApi/Controllers/ProviderController.cs [hunk 22]
```diff
@@ -0,0 +1,145 @@
...
+        public async Task<ActionResult<Provider>> Create(ProviderDto providerDTO)
+        {
+            if (!ModelState.IsValid)
+            {
+                return BadRequest(ModelState);
+            }
+
+            if (providerDTO == null)
+            {
+                return BadRequest(providerDTO);
```
Comment (dmitrykiev) [inline OutOfSchool/OutOfSchool.WebApi/Controllers/ProviderController.cs:83]: does any error message needed here?

File: OutOfSchool/OutOfSchool.WebApi/Controllers/ProviderController.cs [hunk 23]
```diff
@@ -0,0 +1,145 @@
...
+    [Authorize(AuthenticationSchemes = "Bearer")]
+    public class ProviderController : ControllerBase
+    {
+        private readonly ILogger<ProviderController> logger;
+        private readonly IProviderService providerService;
+
+        /// <summary>
+        /// Initializes a new instance of the <see cref="ProviderController"/> class.
+        /// </summary>
+        /// <param name="logger">Logging mechanism.</param>
```
Comment (dmitrykiev) [inline OutOfSchool/OutOfSchool.WebApi/Controllers/ProviderController.cs:25]: maybe "Logger instance" sounds better

File: OutOfSchool/OutOfSchool.WebApi/Controllers/ProviderController.cs (hunk 21 above)
Comment (VyacheslavDzhus) [reply OutOfSchool/OutOfSchool.WebApi/Controllers/ProviderController.cs:120]: He said to use CreatedAtAction with Post method, here we have the Put method

File: OutOfSchool/OutOfSchool.WebApi/Controllers/ProviderController.cs (hunk 22 above)
Comment (VyacheslavDzhus) [reply OutOfSchool/OutOfSchool.WebApi/Controllers/ProviderController.cs:83]: StyleCop asks for null check

Comment (dmitrykiev) [reply OutOfSchool/OutOfSchool.WebApi/Controllers/ProviderController.cs:83]: will user understand what went wrong in this case?

File: OutOfSchool/OutOfSchool.DataAccess/Models/Provider.cs [hunk 24]
```diff
@@ -0,0 +1,92 @@
...
+
+namespace OutOfSchool.Services.Models
+{
+    public class Provider
+    {
+        public long Id { get; set; }
+
+        [Required(ErrorMessage = "Title is required")]
+        [DataType(DataType.Text)]
+        public string Title { get; set; }
```
Comment (mmmpolishchuk) [inline OutOfSchool/OutOfSchool.DataAccess/Models/Provider.cs:16]: Shouldn't it be like this: public string Title { get; set; } = string.Empty ?

Comment (VyacheslavDzhus) [reply OutOfSchool/OutOfSchool.DataAccess/Models/Provider.cs:16]: We decided not to use this for properties with attribute Required

**Review by dmitrykiev — APPROVED** [review]
